### DISCUSSION TITLE
FullFuzz test exploits-minimized

### DISCUSSION THREAD

**Discussion by guyzilberblum** [Cyfrin/security-and-auditing-full-course-s23#176] [Q&A]
While im watching the tutuorial about the full fuzz test patrick did forge test and  broke the invariant at the do more math function 
while i did as he showed and the test passed without finding any bug notice that i did try to switch the fail on revert to false this is my code 
""" 

        // SPDX-License-Identifier: MIT
            pragma solidity 0.8.20;
            import {Test} from "forge-std/Test.sol";
            import{StatefulFuzzCatches} from "../../src/invariant-break/StatefulFuzzCatches.sol";
            import {StdInvariant} from "forge-std/StdInvariant.sol";
            contract StateFUllfuzz is StdInvariant,Test {
           StatefulFuzzCatches sff;
           function setUp() public {
           sff = new StatefulFuzzCatches();
          targetContract(address(sff));
          }

        function testFuzzStatelessDoMoreMath (uint128 randNumber) public  {
        assert(sff.doMoreMathAgain(randNumber)!=0);

        }

       function testStatefulfuzz_CatchesInvariant() public view {
       
        assert(sff.storedValue()!= 0);
         
      }
      }

"""

**Comment by EngrPips** [comment]
```js
 // SPDX-License-Identifier: MIT
        pragma solidity 0.8.20;
        import {Test} from "forge-std/Test.sol";
        import{StatefulFuzzCatches} from "../../src/invariant-break/StatefulFuzzCatches.sol";
        import {StdInvariant} from "forge-std/StdInvariant.sol";
        contract StateFUllfuzz is StdInvariant,Test {
       StatefulFuzzCatches sff;
       function setUp() public {
       sff = new StatefulFuzzCatches();
      targetContract(address(sff));
      }


    function testFuzzStatelessDoMoreMath (uint128 randNumber) public  {
    assert(sff.doMoreMathAgain(randNumber)!=0);

    }


   function testStatefulfuzz_CatchesInvariant() public view {
   
    assert(sff.storedValue()!= 0);
     
  }
  }
```

**Reply by EngrPips** [reply]
Hello @guyzilberblum, Fuzz testing strongly depends on the seed used and the number of runs made to determine the outcome. So, if you would like to have the same output as Patrick, you need to be running your fuzz testing with the same seed he is using and the same runs he has set in his `foundry.toml`

**Reply by kapitanluffy5** [reply]
This what I got ![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/152165691/07417bcc-e8f3-4e74-8ed4-699c85e41569)

**Reply by kapitanluffy5** [reply]
This is what Patrick got ![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/152165691/8456d690-dfe0-4d6b-82db-a75ad7abfe6d)

**Reply by EngrPips** [reply]
You are running a `unit test` while Patrick ran a `Fuzz test.` That is because you named the function that is supposed to run the `Fuzz test` as `testStatefulFuzz_CatchesInvariant` when, in fact, you are supposed to name it `statefulFuzz_CatchesInvariant.`

**Comment by guyzilberblum** [comment]
I used the same stats

**Reply by EngrPips** [reply]
You are supposed to get the same output if you use the same stats. That's weird; you can share your test and output and Patrick's test and output so we can see what is different.

**Comment by guyzilberblum** [comment]
Ok I will send tomorrow it’s the weekend here I’m going out thanks a lot
talk soon

**Reply by EngrPips** [reply]
Alright, friend. Enjoy your weekend, and I will be expecting those

**Comment by guyzilberblum** [comment]
Thanks, you too!

**Reply by EngrPips** [reply]
Yeah, sure I will, hahaha

**Comment by kapitanluffy5** [comment]
No it did more tests but still the test passed no errors

**Reply by EngrPips** [reply]
I am not sure how you mean.

**Comment by guyzilberblum** [comment]
It still doesn't work

**Reply by EngrPips** [reply]
Show your test and the output again.

**Reply by kapitanluffy5** [reply]
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/152165691/b34a7b0c-dce2-479b-a287-28bca021a0ee)

**Reply by EngrPips** [reply]
I am trying to understand why your `fuzz` test doesn't show details like Patrick `fuzz` test did. I mean the `(runs: , calls: , reverts: )`

**Comment by kapitanluffy5** [comment]
I have no idea I will sit over the code again tomorrow gm

**Reply by EngrPips** [reply]
Alright, friend. Let me know whatever you find.

**Comment by kapitanluffy5** [comment]
Ok I changed the seed to 0x5 and it find an assertion failed I don’t know
if it’s like assertion violation but it solved the problem

**Reply by EngrPips** [reply]
Yeah, that is the same. Awesome!

**Reply by sharonphiliplima** [reply]
Help me out too please

```
[fuzz]
runs = 1000
seed = '0x5' 

[invariant]
runs = 256
depth = 15
fail_on_revert = false
```

and 
```
contract StatefulFuzzCatchesTest is StdInvariant, Test {

    StatefulFuzzCatches statefulFuzzCatches;
    function setUp() public{
        statefulFuzzCatches = new StatefulFuzzCatches();
        targetContract(address(statefulFuzzCatches));
    }

    function testDoMoreMathAgain(uint128 randomNumber) public {
        assert(statefulFuzzCatches.doMoreMathAgain(randomNumber) != 0);
    }

    function statefulFuzz_catchesInvariant() public view{
        assert(statefulFuzzCatches.storedValue() != 0);

    }
}

```
```
forge test --mt statefulFuzz_catchesInvariant
[⠒] Compiling...
No files changed, compilation skipped

Ran 1 test for test/invariant-break/StatefulFuzzCatchesTest.t.sol:StatefulFuzzCatchesTest
[PASS] statefulFuzz_catchesInvariant() (runs: 256, calls: 3840, reverts: 49)
Suite result: ok. 1 passed; 0 failed; 0 skipped; finished in 490.99ms (482.83ms CPU time)

Ran 1 test suite in 494.56ms (490.99ms CPU time): 1 tests passed, 0 failed, 0 skipped (1 total tests)

```

**Reply by EngrPips** [reply]
Hey, You should increase the `invariant` runs, and maybe it will catch that scenario where `storedValue` is `0.`

**Reply by sharonphiliplima** [reply]
Got it. So, I had to actually make the runs = 1000 to get it work! Thanks.

**Reply by sharonphiliplima** [reply]
Does that mean that we can't rely on it?

**Reply by EngrPips** [reply]
Actually, fuzzing isn't 100% reliable because some edge cases can be too far away in the number of runs needed to catch them, and that is why we have `formal verification,` which does not throw random value to a contract to break it but instead prove that a condition holds mathematically. You know mathematics never lied, lol. When you get to the `evm and formal verification` course of the Cyfrin curriculum, you will learn in great detail `formal-verification.`
